File: 0543-diameter-of-binary-tree/solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int diameter = 0;
int getHeight(struct TreeNode* root){
    if( root == NULL) return 0;
    int lw = getHeight(root->left);
    int rw = getHeight(root->right);
    diameter = fmax(diameter,lw+rw);
    return fmax(lw,rw) + 1;
}
int diameterOfBinaryTree(struct TreeNode* root) {
    diameter = 0;
    getHeight(root);
    return diameter; 
}

File: 0543-diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int ans = 0;
    private int height(TreeNode node){
        if(node == null) return 0;
        int left = height(node.left);
        int right = height(node.right);
        ans = Math.max(ans,left+right);
        return Math.max(left,right) + 1;
    }
    public int diameterOfBinaryTree(TreeNode root) {
       height(root);
       return ans;
    }
}
